metricbeat/beater/reload.go

```diff
@@ -0,0 +1,221 @@
+package beater
+
+import (
+	"expvar"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/publisher"
+	"github.com/elastic/beats/metricbeat/mb"
+)
+
+var (
+	debugr        = logp.MakeDebug("reloader")
+	configReloads = expvar.NewInt("metricbeat.config.reloads")
+	moduleStarts  = expvar.NewInt("metricbeat.config.module.starts")
+	moduleStops   = expvar.NewInt("metricbeat.config.module.stops")
+	moduleRunning = expvar.NewInt("metricbeat.config.module.running")
+
+	ErrNoNewFiles = errors.New("no new files found")
+)
+
+type ConfigReloader struct {
+	registry RunningRegistry
+	config   ModulesReloadConfig
+	client   func() publisher.Client
+	done     chan struct{}
+	wg       sync.WaitGroup
+}
+
+type RunningRegistry struct {
+	List  map[uint64]ModuleRunner
+	mutex sync.Mutex
+}
+
+func NewConfigReloader(config ModulesReloadConfig, p publisher.Publisher) *ConfigReloader {
+
+	client := func() publisher.Client { return p.Connect() }
+
+	return &ConfigReloader{
+		registry: NewRunningRegistry(),
+		config:   config,
+		client:   client,
+		done:     make(chan struct{}),
+	}
+}
+
+func (r *ConfigReloader) Run() {
+
+	logp.Info("Config reloader started")
+
+	r.wg.Add(1)
+	defer r.wg.Done()
+
+	// Stop all running modules when method finishes
+	defer r.StopModules(r.registry.CopyList())
+
+	gcd := NewGlobalChangeDetector(r.config.Path)
+
+	for {
+		select {
+		case <-r.done:
+			logp.Info("Dynamic config reloader stopped")
+			return
+		case <-time.After(r.config.Period):
+
+			debugr("Scan for new config files")
+			configReloads.Add(1)
+
+			files, updated, err := gcd.Scan()
+			if err != nil {
+				logp.Err("Error fetching new config files: %v", err)
+				// Continue on error or no new files
+				continue
+			}
+
+			// no file changes
+			if !updated {
+				continue
+			}
+
+			// Load all config objects
+			configs := []*common.Config{}
+			for _, file := range files {
+				c, err := LoadConfigs(file)
+				if err != nil {
+					logp.Err("Error loading config: %s", err)
+					continue
+				}
+
+				configs = append(configs, c...)
+			}
+
+			// Check which configs do not exist anymore
+			s, err := NewModuleWrappers(configs, mb.Registry)
+			if err != nil {
+				if err != mb.ErrAllModulesDisabled && err != mb.ErrEmptyConfig {
+					// Continuing as only some modules could have an error
+					logp.Err("Error creating modules: %s", err)
+				}
+			}
+
+			debugr("Number of module wrappers created: %v", len(s))
+
+			startList := []*ModuleWrapper{}
```

Comment:
Use `var startList []*ModuleWrapper` to avoid allocating an empty map if it doesn't get used. [ref](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices)